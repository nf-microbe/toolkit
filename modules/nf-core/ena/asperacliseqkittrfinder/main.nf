process ENA_ASPERACLISEQKITTRFINDER {
    tag "${meta.id}"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'oras://community.wave.seqera.io/library/aspera-cli_seqkit_biopython:b1c05ee4816a116b' :
        'community.wave.seqera.io/library/aspera-cli_seqkit_biopython:ce3fa7446f366f82' }"

    input:
    tuple val(meta), val(url)

    output:
    tuple val(meta), path("${prefix}.trfinder.fasta.gz")    , emit: fasta
    tuple val(meta), path("${prefix}.trfinder.tsv")         , emit: tsv
    path "versions.yml"                                     , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def url_list    = url.collect { urls -> urls.toString() }.join(',')
    def args        = task.ext.args ?: ''
    def args2       = task.ext.args2 ?: ''
    def conda_prefix= ['singularity', 'apptainer'].contains(workflow.containerEngine) ? "export CONDA_PREFIX=/opt/conda" : ""
    prefix          = task.ext.prefix ?: "${meta.id}"
    """
    mkdir -p tmp/download tmp/seqkit tmp/trfinder
    ${conda_prefix}
    IFS=',' read -r -a url_array <<< "${url_list}"

    ### Download ENA assemblies
    printf '%s\\n' "\${url_array[@]}" | xargs -I{} -n 1 -P ${task.cpus} bash -c \\
    'IFS=":" read -ra file <<< "{}"; \\
    mod_file=\$(echo \${file[1]:1} | tr / _); \\
    echo \${mod_file}; \\
    ascp \\
        -QT -l 300m -P 33001 \\
        -i \$CONDA_PREFIX/etc/aspera/aspera_bypass_dsa.pem \\
        era-fasp@{} \\
        tmp/download/\$mod_file.fna.gz'

    ### Remove short contigs
    for file in tmp/download/*; do
        filename=\$(basename \$file)

        seqkit \\
            seq \\
            --threads ${task.cpus} \\
            ${args} \\
            \$file \\
            --out-file tmp/seqkit/\${filename%%.*}.fasta
    done

    rm -rf tmp/download/

    ### Identify Terminal repeats
    cd tmp/trfinder

    for file in ../seqkit/*.fasta; do
        filename=\$(basename \$file)

        trfinder.py \\
            --input \$file \\
            --prefix ENA_\${filename%.*} \\
            ${args2}
    done

    cd ../..
    rm -rf tmp/seqkit/

    # combine and compress tr finder files
    cat tmp/trfinder/*.trfinder.fasta > ${prefix}.trfinder.fasta
    awk '(NR == 1) || (FNR > 1)' tmp/trfinder/*.trfinder.tsv > ${prefix}.trfinder.tsv

    rm -rf tmp/trfinder/
    gzip -f ${prefix}.trfinder.fasta

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ascli: 4.14.0
        seqkit: \$(seqkit version | cut -d' ' -f2)
        python: \$( python --version | sed 's/Python //' )
        biopython: \$(python -c "import Bio; print(Bio.__version__)")
    END_VERSIONS
    """

    stub:
    def args   = task.ext.args ?: ''
    def args2   = task.ext.args2 ?: ''
    def args3   = task.ext.args3 ?: ''
    prefix  = task.ext.prefix ?: "${meta.id}"
    """
    echo "" | gzip > ${prefix}.trfinder.fasta.gz
    touch ${prefix}.trfinder.tsv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ascli: 4.14.0
        seqkit: \$(seqkit version | cut -d' ' -f2)
        python: \$( python --version | sed 's/Python //' )
        biopython: \$(python -c "import Bio; print(Bio.__version__)")
    END_VERSIONS
    """
}
