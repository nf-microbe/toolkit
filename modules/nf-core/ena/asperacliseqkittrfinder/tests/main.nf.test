nextflow_process {

    name "Test Process ENA_ASPERACLISEQKITTRFINDER"
    script "../main.nf"
    process "ENA_ASPERACLISEQKITTRFINDER"

    tag "modules"
    tag "modules_nfcore"
    tag "ena"
    tag "ena/asperacliseqkittrfinder"

    test("ERZ1023958 + ERZ1024284") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test' ],
                    [
                        "fasp.sra.ebi.ac.uk:/vol1/sequence/ERZ102/ERZ1023958/contig.fa.gz",
                        "fasp.sra.ebi.ac.uk:/vol1/sequence/ERZ102/ERZ1024284/contig.fa.gz"
                    ]
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test' ],
                    [
                        "fasp.sra.ebi.ac.uk:/vol1/sequence/ERZ102/ERZ1023958/contig.fa.gz",
                        "fasp.sra.ebi.ac.uk:/vol1/sequence/ERZ102/ERZ1024284/contig.fa.gz"
                    ]
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
