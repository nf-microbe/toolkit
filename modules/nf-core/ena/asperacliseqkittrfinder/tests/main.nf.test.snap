{
    "stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.trfinder.fasta.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.trfinder.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,87d06848c4261e79111595a35a4ada46"
                ],
                "fasta": [
                    [
                        {
                            "id": "test"
                        },
                        "test.trfinder.fasta.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "tsv": [
                    [
                        {
                            "id": "test"
                        },
                        "test.trfinder.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,87d06848c4261e79111595a35a4ada46"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-11T11:27:17.616073058"
    },
    "ERZ1023958 + ERZ1024284": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.trfinder.fasta.gz:md5,5d813f97258f7786b8dff1e347fff1ae"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.trfinder.tsv:md5,96d620739c1fcac7b38829381fdd043e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,87d06848c4261e79111595a35a4ada46"
                ],
                "fasta": [
                    [
                        {
                            "id": "test"
                        },
                        "test.trfinder.fasta.gz:md5,5d813f97258f7786b8dff1e347fff1ae"
                    ]
                ],
                "tsv": [
                    [
                        {
                            "id": "test"
                        },
                        "test.trfinder.tsv:md5,96d620739c1fcac7b38829381fdd043e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,87d06848c4261e79111595a35a4ada46"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-11T15:32:34.00834446"
    }
}